Adding SlotFill support for API-driven tasks components

joelclimbsthings opened the pull request:
Fixes #7373

This PR implements SlotFill for optional use in replacing a task's appearance within a task list, as well as utilizing it for displaying all current task components once clicked on. 

Adds slot-fill components to onboarding package:
- `<WooOnboardingTaskListItem />`
- `<WooOnboardingTask />`

_Notes:_
- This involves copying over all the current task components and temporarily duplicating them (while modifying the new ones). This creates a less-than-pretty diff, but the copy/paste changes are captured in the single commit d7b1c4f. It might be best to parse through the other commits instead of looking at the global diff.
- A few things, such as body classes and other classes for styles seem a little messy. They were already used elsewhere in this feature so this might be a target of a later clean-up PR. 
- I had to move around the new logic regarding `trackedCompletedActions` for the Marketing task in commit ffa9d85d9ec7b8f5db00ab808450f83caabe9f91. @louwie17 might want to ensure this is as expected.

### Detailed test instructions:

-   Checkout branch
-   Ensure tasks feature is enabled
- Smoke test all tasks to ensure that they appear and function as they have previously
- You can test the `WooOnboardingTaskListItem` by creating and importing a test component that looks something like below. This will result in the title for the payments task changing to "Alternative payments task title."
```js
/**
 * External dependencies
 */
import { registerPlugin } from '@wordpress/plugins';
import { WooOnboardingTaskListItem } from '@woocommerce/onboarding';

registerPlugin( 'wc-admin-onboarding-slot-test', {
	scope: 'woocommerce-admin',
	render: () => (
		<WooOnboardingTaskListItem id="payments">
			{ ( { defaultTaskItem } ) => (
				<>
					{ React.cloneElement( defaultTaskItem, {
						title: 'Alternative payments task title',
					} ) }
				</>
			) }
		</WooOnboardingTaskListItem>
	),
} );
```

joelclimbsthings committed:
Basics for WooOnboardingTaskListItem Slotfill

joelclimbsthings committed:
Copying tasks over and wiring up the first two

joelclimbsthings committed:
Migrating marketing task

joelclimbsthings committed:
Migrating remaining tasks

joelclimbsthings committed:
Adding style dependent classes for tasks

joelclimbsthings committed:
Adding changelog

joshuatf commented on client/tasks/tasks-fills/Marketing/index.tsx:
```diff
@@ -0,0 +1,259 @@
+	scope: 'woocommerce-admin',
+	render: () => (
+		<WooOnboardingTask id="marketing">
+			<Marketing />;
```
```suggestion
			<Marketing />
```

joshuatf commented on client/tasks/tasks-fills/shipping/index.js:
```diff
@@ -0,0 +1,383 @@
+	scope: 'woocommerce-admin',
+	render: () => (
+		<WooOnboardingTask id="shipping">
+			<ShippingWrapper />;
```
```suggestion
			<ShippingWrapper />
```

joshuatf commented on client/tasks/index.ts:
```diff
@@ -2,6 +2,7 @@
  * Internal dependencies
  */
 import { Tasks } from './tasks';
+import './tasks-fills';
```
Nitpicky, but can we name this `fills` to keep with the simplified subdirectory naming scheme?

joshuatf reviewed (commented):
Great work, Joel!  This is testing really well for me so far.  Left a couple suggestions for straggler semicolons and also noticing a couple missing tasks.

* Purchase - Also the `actionUrl` from the backend needs updating, but this will probably require a rebase
* WC Pay

joelclimbsthings committed:
Removing straggler semi-colons

joelclimbsthings committed:
Renaming task-fills directory to fills

joelclimbsthings committed:
Adding wcpay task slot fill

joelclimbsthings committed:
Adding purchase task

joelclimbsthings committed:
Changing plugin scope to woocommerce-tasks

joelclimbsthings commented:
Thanks much for the review @joshuatf ! Agree with your comments and I believe I've addressed it all.

>  Left a couple suggestions for straggler semicolons and also noticing a couple missing tasks.

Those were a bit more fun than the others to piece together; should be working now!

joshuatf commented on client/tasks/fills/purchase.js:
```diff
@@ -0,0 +1,73 @@
+				<>
+					{ React.cloneElement( defaultTaskItem, {
+						onClick: () => {
+							onClickActions();
```
Is there any way we can handle these onclick events in the parent?  I'm concerned consumers may forget to implement and we lose tracking.

joelclimbsthings replied:
Yeah, something that I'd been considered considering as well. I'll look into this real quick.

joelclimbsthings replied:
I've reworked this in 64841e097b6d439067ac7fa595addc1d55e668a6.

joshuatf commented on client/tasks/fills/woocommerce-payments.js:
```diff
@@ -0,0 +1,45 @@
+						onClick: () => {
+							onClickActions();
+
+							return new Promise( ( resolve, reject ) => {
```
I'm not sure if it's still handled this way in the experimental task list item, but previously the `onClick` would accept a promise and automatically have a busy state until it was resolved.

If that's the case, something may be broken here.  And if not, we should probably add some indication of busy state while this is installing/activating, but also okay if we want to handle in a follow-up.

joelclimbsthings replied:
I don't see any logic capturing a promise via the onClick handler. I _have_ noticed that some of the buttons elsewhere on the page will display a busy state whenever `installPlugins` is being requested in the plugins store, which can be triggered by clicking this task.

The busy states I've seen are triggered by buttons, and I'm not certain how this might appear when triggered by a task? @joshuatf

joshuatf commented on client/tasks/fills/purchase.js:
```diff
@@ -0,0 +1,73 @@
+/**
+ * Internal dependencies
+ */
+import CartModal from '../../dashboard/components/cart-modal';
```
I believe pre-existing, but just noting some minor styling issues with the modal spacing:
![Screen Shot 2021-09-17 at 3 06 56 PM](https://user-images.githubusercontent.com/10561050/133843640-96a3341b-a893-458d-a460-66949c0aac02.png)

joelclimbsthings replied:
Nice catch, I've gone ahead and addressed the gap under the heading in 64841e097b6d439067ac7fa595addc1d55e668a6.

joshuatf reviewed (commented):
Excellent work, @joelclimbsthings!

Left a few more thoughts on this PR, but let me know if you think they're more appropriate for a follow-up.

I'd love to get some more 👀  on this one, but I think we can issue a call for testing prior to turning this feature flag on.

joelclimbsthings committed:
Refactoring onClick handling for task item slotfill to ensure tracking always executes

joelclimbsthings committed:
Making gap in cart modal larger

joelclimbsthings commented:
Thanks for the review @joshuatf . I've addressed the feedback, with a little uncertainty around how to address the busy state you mentioned. See my other [other comment](https://github.com/woocommerce/woocommerce-admin/pull/7616#discussion_r712570285) regarding that.

joshuatf commented on packages/experimental/src/experimental-list/task-item.tsx:
```diff
@@ -43,8 +43,8 @@ type TaskItemProps = {
 	onCollapse?: () => void;
 	onDelete?: () => void;
 	onDismiss?: () => void;
+	onSnooze?: () => void;
```
Let's add a changelog entry for this under `packages/experimental`.

joelclimbsthings replied:
Done 👍🏻

joshuatf commented on changelogs/add-7373-slotfill-task-api:
```diff
@@ -0,0 +1,4 @@
+Significance: minor
+Type: Add
+
+Adding Slotfill support to new API-driven tasks components
```
I think it might be a good idea to have a single changelog entry for the new task API features.  I don't have a strong opinion on this, but my feeling is that this could be a confusing entry when the feature is not yet turned on.

joelclimbsthings replied:
True, glad to remove this here.

joshuatf commented on client/tasks/tasks.tsx:
```diff
@@ -70,6 +70,11 @@ export const Tasks: React.FC< TasksProps > = ( { query } ) => {
 
+	useEffect( () => {
+		document.body.classList.add( 'woocommerce-onboarding' );
+		document.body.classList.add( 'woocommerce-task-dashboard__body' );
```
I know this is just migrating old code, but it seems like this never gets removed from the page when navigating away.  Not entirely sure why we need this body classes at all actually.

Okay to handle in a follow-up though- would you mind creating an issue if we don't remove here?

joelclimbsthings replied:
Yeah @joshuatf , I actually toyed around with removing these earlier, but some of the styles for the various tasks depend on these body classes. I actually mentioned this in the notes above ^. I'm good with creating an issue for a follow-up 👍🏻

joelclimbsthings replied:
[Done](https://github.com/woocommerce/woocommerce-admin/issues/7693)

joshuatf commented on client/tasks/task-list-item.tsx:
```diff
@@ -112,18 +112,18 @@ export const TaskListItem: React.FC< TaskListItemProps > = ( {
 	};
 
-	const onClickActions = useCallback( () => {
+	const onClickTracking = () => {
```
Nitpicky, but maybe `trackClick` is more semantic?

joshuatf commented on client/tasks/task-list-item.tsx:
```diff
@@ -146,34 +146,47 @@ export const TaskListItem: React.FC< TaskListItemProps > = ( {
 		/>
 	) : (
-		defaultTaskItem
+		<DefaultTaskItem />
```
Nice 💯

joshuatf reviewed (approved):
Great work, @joelclimbsthings!  Left a few nitpicky comments, but nothing major and pre-approving pending the fix of failing tests.

>That said, the busy states I've seen are triggered by buttons, and I'm not certain how this might appear when triggered by a task? @joshuatf

I don't see this logic anymore either, so I'm wondering if it got removed.  I'll create a follow-up issue so we can address.

joelclimbsthings committed:
Renaming functions to be a bit more semantic

joelclimbsthings committed:
Removing changelog

joelclimbsthings committed:
Adding changloge to experimental package
